### PR TITLE
[proto] Ported RandomIoUCrop from detection refs

### PR DESCRIPTION
**Pull request by vfdev-5**
- [x] Ported RandomIoUCrop transform
- [x] Check visually
- [x] Tests to add

### CODE REVIEW

File: torchvision/prototype/transforms/_geometry.py
Comment (pmeier) [inline torchvision/prototype/transforms/_geometry.py:697]: This shows one of the issues I highlighted in the design doc: here we assume that every `Label` we have in the pytree is "linked" to the bounding boxes, i.e. has the same batch dimensions. What if have another label in there for the whole image or this assumption is otherwise violated, e.g. the labels are not a batched tensor but rather a list of labels?

In our detection references we can make this assumption, because we control the input. I have no idea how we want to handle this in a general way and this is why I haven't ported the remaining detection transforms yet.

Comment (datumbox) [reply torchvision/prototype/transforms/_geometry.py:697]: I think you can safely make the assumption that the labels are linked to the bboxes, for all the transforms under references/detection. Would that be enough to unblock you?

File: torchvision/prototype/transforms/_geometry.py
Comment (vfdev-5) [inline torchvision/prototype/transforms/_geometry.py:713]: @datumbox you were previously saying about filtering input data types. How about this one ?

Comment (datumbox) [reply torchvision/prototype/transforms/_geometry.py:713]: Not 100% sure why we can't support it or whether we should. Perhaps OneHotLabel should inherit from Label so that we don't have to check both of them in `has_all()` calls.

File: torchvision/prototype/transforms/_geometry.py
Comment (pmeier) [inline torchvision/prototype/transforms/_geometry.py:711]: Do we want to enforce labels here? This transform works fine if we don't have labels, right?

File: torchvision/prototype/transforms/functional/_meta.py
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_meta.py:64]: Can we maybe also make this a method on `features.BoundingBox`? I'm aware that `clamp` is already used and so is `clip`. Maybe we can call it `trim`?

File: torchvision/prototype/transforms/_geometry.py
Comment (datumbox) [reply torchvision/prototype/transforms/_geometry.py:711]: We do. Object Detection needs not only BBoxes but also Labels to work.

File: torchvision/prototype/transforms/functional/_meta.py
Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_meta.py:64]: We may try to override existing `Tensor.clamp` ?

Comment (datumbox) [reply torchvision/prototype/transforms/functional/_meta.py:64]: > We may try to override existing Tensor.clamp ?

I would avoid overwriting default tensor methods. Even deprecated ones might be a problem.

Comment (vfdev-5) [reply torchvision/prototype/transforms/functional/_meta.py:64]: Why do you think it could be harmful ?

Comment (datumbox) [reply torchvision/prototype/transforms/functional/_meta.py:64]: I know we do this more on the resize which is deprecated but as we discussed in that case it might surprise users. We should probably ask for feedback from the community on the topic. From one hand, subclassing Tensor means we can overwrite methods to make the more "relevant" to the specific type. On the other hand, it might confuse users who expect that they are accessing the parent method. To avoid blocking this PR, I think we can go ahead without a method on the BBox object and raise this as an open question on the API design.

File: torchvision/prototype/transforms/_geometry.py
Comment (datumbox) [inline torchvision/prototype/transforms/_geometry.py:703]: Should we be supporting here OneHotLabels too explicitly until #6420 is resolved?

Comment (vfdev-5) [reply torchvision/prototype/transforms/_geometry.py:703]: Yes, we should. I have to understand how it would work with OneHotLabels in this case...
I do not plan to continue on #6420 right now

File: torchvision/prototype/transforms/_geometry.py
Comment (datumbox) [inline torchvision/prototype/transforms/_geometry.py:710]: It feels that this clamping operation should have been also done on all other Crop ops. Do you plan to add them in this PR or on separate one?

File: torchvision/prototype/transforms/_geometry.py
Comment (datumbox) [inline torchvision/prototype/transforms/_geometry.py:727]: We should also add a quick support of Masks. The previous operator didn't support it on references but now we can do this. We just need to crop it exactly how we do the image and we should be OK.

Comment (vfdev-5) [reply torchvision/prototype/transforms/_geometry.py:727]: Masks are implicitly supported. I added a test case

Comment (datumbox) [reply torchvision/prototype/transforms/_geometry.py:727]: Can you update the error message?

**Review by datumbox — APPROVED** [review]
LGTM.

@pmeier any thoughts?

File: torchvision/prototype/transforms/_utils.py
Comment (pmeier) [inline torchvision/prototype/transforms/_utils.py:20]: 1. We are not querying boxes, since we only return one.
2. Let's keep the naming scheme consistent with the kernels.

```suggestion
def query_bounding_box(sample: Any) -> features.BoundingBox:
```

File: torchvision/prototype/transforms/_utils.py
Comment (pmeier) [inline torchvision/prototype/transforms/_utils.py:26]:
```suggestion
    raise TypeError("No bounding box was found in the sample")
```

**Review by pmeier — COMMENTED** [review]
Small naming nit for consistency. Otherwise LGTM.

File: torchvision/prototype/transforms/_geometry.py
Comment (vfdev-5) [inline torchvision/prototype/transforms/_geometry.py:716]: @datumbox here is a support for one-hot encoded masks meanwhile other solutions we could decide about

File: torchvision/prototype/transforms/_geometry.py
Comment (pmeier) [inline torchvision/prototype/transforms/_geometry.py:724]: What about plain tensors? For `CutMix` and `MixUp` we don't allow the "old" image types at all:

https://github.com/pytorch/vision/blob/162267ca90a8a7d03a8d9249602d6c72b497656a/torchvision/prototype/transforms/_augment.py#L108-L109

Given that have ported this from references there is no BC constraint to allow plain tensors and PIL images. Still, it feels unnecessary restrictive.

File: torchvision/prototype/transforms/_geometry.py
Comment (pmeier) [inline torchvision/prototype/transforms/_geometry.py:731]:
```suggestion
        return super().forward(sample)
```

File: torchvision/prototype/transforms/_geometry.py
Comment (vfdev-5) [reply torchvision/prototype/transforms/_geometry.py:724]: To understand your comment, you want to add support for images as Tensors and keep Image and PIL Image ?

Comment (pmeier) [reply torchvision/prototype/transforms/_geometry.py:724]: Either that or remove support for PIL here. We should either support all image types or only the "new one" like we do in `CutMix` and `MixUp` (not saying this is a good thing, but we should be consistent).

Comment (vfdev-5) [reply torchvision/prototype/transforms/_geometry.py:724]: Is the reason why CutMix, MixUp are not supporting PIL due to lack of implementation ?
Here in RandomIoUCrop we can support everything, so I would add image as torch.Tensor support and keep PIL.

Comment (datumbox) [reply torchvision/prototype/transforms/_geometry.py:724]: I think we should support all types. Part of the API supporting them and part not is weird and will hinter adoption.

**Pull request merged**
